ssl/quic/quic_ackm.c

```diff
@@ -0,0 +1,1997 @@
+ * set at all, this is a no-op and returns 1.
+ */
+static int pn_set_remove(struct pn_set_st *s, const OSSL_QUIC_ACK_RANGE *range)
+{
+    struct pn_set_item_st *z, *zprev, *y;
+    QUIC_PN start = range->start, end = range->end;
+
+    if (!ossl_assert(start <= end))
+        return 0;
+
+    /* Walk backwards since we will most often be removing at the end. */
+    for (z = s->tail; z != NULL; z = zprev) {
+        zprev = z->prev;
+
+        if (z->range.end < start)
+            /* No overlapping ranges can exist beyond this point, so stop. */
+            break;
+
+        if (start <= z->range.start && end >= z->range.end) {
+            /*
+             * The range being removed dwarfs this range, so it should be
+             * removed.
+             */
+            if (z->next != NULL)
+                z->next->prev = z->prev;
+            if (z->prev != NULL)
+                z->prev->next = z->next;
+            if (s->head == z)
+                s->head = z->next;
+            if (s->tail == z)
+                s->tail = z->prev;
+
+            OPENSSL_free(z);
+            --s->num_ranges;
+        } else if (z->range.start >= start) {
+            /*
+             * The range being removed includes start of this range, but does
+             * not cover the entire range (as this would be caught by the case
+             * above). Shorten the range.
+             */
+            assert(end < z->range.end);
+            z->range.start = end + 1;
+        } else if (z->range.end <= end) {
+            /*
+             * The range being removed includes the end of this range, but does
+             * not cover the entire range (as this would be caught by the case
+             * above). Shorten the range. We can also stop iterating.
+             */
+            assert(start > z->range.start);
+            assert(start > 0);
+            z->range.end = start - 1;
+            break;
+        } else if (z->range.start <= start && z->range.end >= end) {
+            /*
+             * The range being removed falls entirely in this range, so cut it
+             * into two. Cases where a zero-length range would be created are
+             * handled by the above cases.
+             */
+            assert(z->range.start < start);
+            assert(z->range.end   > end);
+
+            y = OPENSSL_zalloc(sizeof(struct pn_set_item_st));
+            if (y == NULL)
+                return 0;
+
+            y->range.end   = z->range.end;
+            y->range.start = end + 1;
+            y->next = z->next;
+            y->prev = z;
+            if (y->next != NULL)
+                y->next->prev = y;
+
+            z->range.end = start - 1;
+            z->next = y;
+
+            if (s->tail == z)
+                s->tail = y;
+
+            ++s->num_ranges;
+            break;
+        } else if (pn_range_overlaps(&z->range, range)) {
+            /* Partial overlap. All cases should be covered above. */
+            assert(0);
```

Comment:
Perhaps better would be:

```` C
} else {
    /* Partial overlap. All cases should be covered above. */
    assert(!pn_range_overlaps(&z->range, range));
}
````

This gives a better assert message in the event that the assertion fails and removes the call to `pn_range_overlaps` in a production build.